mppe/crawler.go

```diff
@@ -0,0 +1,257 @@
+			category: "verbas-indenizatorias-e-outras-remuneracoes-temporarias",
+			yearCodes: map[int]int{
+				2018: 415,
+				2019: 451,
+				2020: 510,
+			},
+		},
+	}
+)
+
+// Crawl download all files related to the MPPE salaries and return their local paths
+func Crawl(outputPath string, month, year int) ([]string, error) {
+	paths := make([]string, 8)
+	pathChannel := make(chan string, 8)
+	var wg sync.WaitGroup
+	for _, member := range members {
+		wg.Add(1)
+		go func(member employeeDescriptor, month, year int) {
+			defer wg.Done()
+			link := fmt.Sprintf("%s%d-%s", baseURL, member.yearCodes[year], member.category)
+			resp, err := http.Get(link)
+			if err != nil {
+				return
+			}
+			defer resp.Body.Close()
+			b, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				return
+			}
+			htmlAsString := string(b)
+			pattern := pathResolver(month, year, member.category)
+			fileCode, err := findFileIdentifier(htmlAsString, pattern)
+			if err != nil {
+				return
+			}
+			urlToDownload := fmt.Sprintf("%s%d-%s?download=%s", baseURL, member.yearCodes[year], member.category, fileCode)
+			filePath := fmt.Sprintf("%s/%s-%s-%d.xlsx", outputPath, member.category, fmt.Sprintf("%02d", month), year)
+			desiredFile, err := os.Create(filePath)
+			if err != nil {
+				return
+			}
+			defer desiredFile.Close()
+			err = donwloadFile(urlToDownload, desiredFile)
+			if err != nil {
+				return
+			}
+			pathChannel <- filePath
+		}(member, month, year)
+	}
+	go func() {
+		wg.Wait()
+		close(pathChannel)
+	}()
+	for path := range pathChannel {
+		paths = append(paths, path)
+	}
+	return paths, nil
+}
+
+// it gets a HTML file as a string and searchs inside of it a pattern
+// with only numbers
+func findFileIdentifier(htmlAsString, pattern string) (string, error) {
+	indexOfPattern := strings.Index(htmlAsString, pattern)
+	nPreviousChars, err := strconv.Atoi(os.Getenv("PREVIOUS_N_CHARS"))
+	if err != nil {
+		nPreviousChars = 10
+	}
+	if indexOfPattern > 0 {
+		substringWithFileIdentifier := htmlAsString[indexOfPattern-nPreviousChars : indexOfPattern]
+		possibleMatches := regexp.MustCompile("[0-9]+").FindAllString(substringWithFileIdentifier, -1)
+		if len(possibleMatches) == 0 {
+			return "nil", fmt.Errorf("was not possible to get file indetifier")
+		}
+		fileIdentifier := possibleMatches[0]
+		return fileIdentifier, nil
+	}
+	return "nil", fmt.Errorf("was not found anny pattern")
+}
+
+// download a file and writes on the given writer
+func donwloadFile(url string, w io.Writer) error {
+	resp, err := http.Get(url)
+	if err != nil {
+		return fmt.Errorf("error downloading file:%q", err)
+	}
+	defer resp.Body.Close()
+	if _, err := io.Copy(w, resp.Body); err != nil {
+		return fmt.Errorf("error copying response content:%q", err)
+	}
+	return nil
+}
+
+// it returns the proper search path for a given member
+func pathResolver(month, year int, member string) string {
+	if member == "remuneracao-de-todos-os-membros-ativos" {
```

Comment:
Gostei bastante de onde viemos parar! Creio que vai ficar mais limpo se for um bloco `switch`. Go aceita switch com strings